substrate/network-libp2p/src/service.rs

```diff
@@ -865,52 +901,47 @@ fn perform_kademlia_query<T, To, St, C>(shared: Arc<Shared>, transport: T,
 					}
 					None
 				},
-				KadQueryEvent::Finished(out) => Some(out),
+				KadQueryEvent::Finished(_) => Some(()),
 			}
 		)
 		.into_future()
 		.map_err(|(err, _)| err)
-		.map(|(out, _)| out.unwrap())
+		.map(|_| ())
 }
 
-/// Processes the results of a Kademlia discovery.
-fn process_kad_results<T, To, St, C>(shared: Arc<Shared>, transport: T,
-	swarm_controller: SwarmController<St>, results: Vec<PeerstorePeerId>,
-	local_peer_id: &PeerstorePeerId)
+/// Connects to additional nodes, if necessary.
+fn connect_to_nodes<T, To, St, C>(
+	shared: Arc<Shared>,
+	base_transport: T,
+	swarm_controller: &SwarmController<St>
+)
 	where T: MuxedTransport<Output =  TransportOutput<To>> + Clone + 'static,
 		T::MultiaddrFuture: 'static,
 		To: AsyncRead + AsyncWrite + 'static,
 		St: MuxedTransport<Output = (FinalUpgrade<C>, Endpoint)> + Clone + 'static,
 		C: 'static {
-	trace!(target: "sub-libp2p", "Processing Kademlia discovery results \
-		(len = {})", results.len());
-
-	for discovered_peer in results {
-		// Skip if we reach `min_peers`.
-		// Also skip nodes we are already connected to, in order to not connect twice.
-		// TODO: better API in network_state
-		if !shared.network_state.should_open_outgoing_connections() ||
-			discovered_peer == *local_peer_id ||
-			shared.network_state.is_peer_disabled(&discovered_peer)
-		{
-			trace!(target: "sub-libp2p", "Skipping discovered node {:?}", discovered_peer);
-			continue
-		}
+	let num_slots = shared.network_state.should_open_outgoing_custom_connections();
+	debug!(target: "sub-libp2p", "Opening up to {} outgoing connections",
+		num_slots);
+
+	for _ in 0 .. num_slots {
+		// Choose a random peer. We are potentially already connected to
+		// this peer, but this is not a problem as this function is called
+		// regularly.
+		// TODO: is it ^ ?
```

Comment:
Eventually this needs to be reworked to select from the peers that are not currently connected and are not being connected. 
Also, the node could be running with more than a thousands slots, so it should not try to connect all of them at the same time.